### PR TITLE
Testing Container for Back-End (#35, #36)

### LINKED ISSUE #35: Investigate how to properly test micro-services
**Is your feature request related to a problem? Please describe.**
Currently, backend testing is not automated

**Describe the solution you'd like**
Check and integrate framework to automate the testing of micro-services

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

### LINKED ISSUE #36: Implement sample backend ARTEMIS testing workflow
**Is your feature request related to a problem? Please describe.**
Currently we do not have a formal workflow to test ARTEMIS functionality in the backend

**Describe the solution you'd like**
1) Generate test BGP updates

2) Feed them into the monitor module

3) Make sure that rabbitmq handles updates correctly

4) Make sure the updates make it to the db properly

5) If related to hijacks, make sure the correct hijacks are created (detection module) and handled by watching rabbitmq

6) Make sure that the hijacks make it to the db properly

7) Make sure that consequent updates for the same hijack, update the hijack

8) Other possible steps

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by slowr**
### Description of PR

Created an additional container that injects custom bgp updates on the rabbitmq consumers and checks the output of detection and all database inserts/updates.

Closes #35
Closes #36 

What component(s) does this PR affect?

- [x] Back-End (Database, Microservices, Containers, etc)
- [ ] Front-End (Flask, API, etc)
- [ ] Docs
- [x] Build System

Does the PR require changes on other components? If yes, please mark the components:

- [ ] Back-End (Database, Microservices, Containers, etc)
- [ ] Front-End (Flask, API, etc)
- [ ] Docs
- [ ] Build System

### Related Issue

#35 #36 
### Type

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Docs update
- [ ] None of the above

### Checklist:

- [x] I have read the **[contributing guide](https://github.com/FORTH-ICS-INSPIRE/artemis-tool/blob/master/CONTRIBUTING.md)** and my code conforms to the guidelines.
- [x] This change requires a change in the documentation.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: backend/docker-compose.test.yml [hunk 1]
```diff
@@ -0,0 +1,83 @@
+version: '3'
+services:
+    sut:
+        build: ./testing/
+        container_name: testing
```
Comment (vkotronis) [inline backend/docker-compose.test.yml]: why is this not named sut?

File: backend/testing/tester.py [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+                        update_queue,
+                        callbacks=[validate_message],
+                    ),
+                    connection.Consumer(
+                        hijack_db_queue,
+                        callbacks=[validate_message]
+                    )
+
+            ):
+                while curr_idx != send_cnt:
```
Comment (vkotronis) [inline backend/testing/tester.py]: Recheck this logic, I think this remains always True and thus testing is eventually aborted after 4 hours. Unapproving PR for now.

File: backend/core/detection.py [hunk 3]
```diff
@@ -80,14 +80,12 @@ def __init__(self, connection: Connection) -> NoReturn:
...
                 durable=False,
                 delivery_mode=1)
-            self.hijack_exchange.declare()
             self.handled_exchange = Exchange(
                 'handled-update',
                 channel=connection,
                 type='direct',
                 durable=False,
                 delivery_mode=1)
-            self.handled_exchange.declare()
```
Comment (vkotronis) [inline backend/core/detection.py]: why is this deleted?

File: backend/core/detection.py [hunk 4]
```diff
@@ -80,14 +80,12 @@ def __init__(self, connection: Connection) -> NoReturn:
                 type='direct',
                 durable=False,
                 delivery_mode=1)
-            self.hijack_exchange.declare()
```
Comment (vkotronis) [inline backend/core/detection.py]: This is handled by the db, correct?

File: backend/core/detection.py (hunk 3 above)
Comment (slowr) [reply backend/core/detection.py]: we left it on the database side

File: backend/core/detection.py (hunk 4 above)
Comment (slowr) [reply backend/core/detection.py]: same as the other one yes

File: backend/testing/tester.py [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+    with Connection(RABBITMQ_HOST) as connection:
+        pg_amq_bridge.declare(channel=connection.default_channel)
+        bind_queue = process_state_queue(connection.default_channel)
+        bind_queue.declare()
+        processes = 0
+        while True:
+            msg = bind_queue.get()
+            if msg:
+                processes += 1
+                if processes == 13:
```
Comment (vkotronis) [inline backend/testing/tester.py]: why 13? Please create a constant and add a comment.

Comment (slowr) [reply backend/testing/tester.py]: 7 modules initialized and 6 are started afterwards making for 7 inserts and 6 updates (13)

**Comment by vkotronis** [comment]
@slowr also do not forget to update the CHANGELOG with your changes regarding testing (before merging). To avoid multiple commits, please also do the following changes on the CHANGELOG file:

Move "Bumped PyYAML from 3.13 to 4.2b4 in /backend" from "Changed" to "Security"

Move "Bumped PyYAML from 3.13 to 4.2b4 in /frontend" from "Changed" to "Security"

And add the new testing backend features (with sempahore, etc.) to the "Added" section

I would advise documenting somewhere in the wiki how one can add a new test.
You can edit: https://github.com/FORTH-ICS-INSPIRE/artemis/wiki#adding-new-artemis-tests

File: backend/testing/tester.py [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+
+        # queues
+        update_queue = Queue(
+            'detection-testing',
+            exchange=pg_amq_bridge,
+            routing_key='update-insert',
+            durable=False,
+            auto_delete=True,
+            max_priority=1,
+            consumer_arguments={'x-priority': 1})
```
Comment (vkotronis) [inline backend/testing/tester.py]: double-check if priority is correct

File: backend/testing/tester.py [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+
+        hijack_queue = Queue(
+            'hijack-testing',
+            exchange=hijack_exchange,
+            routing_key='update',
+            durable=False,
+            auto_delete=True,
+            max_priority=1,
+            consumer_arguments={'x-priority': 1})
```
Comment (vkotronis) [inline backend/testing/tester.py]: double-check if priority is correct

File: backend/testing/tester.py [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+
+        hijack_db_queue = Queue(
+            'hijack-db-testing',
+            exchange=pg_amq_bridge,
+            routing_key='hijack-update',
+            durable=False,
+            auto_delete=True,
+            max_priority=1,
+            consumer_arguments={'x-priority': 1})
```
Comment (vkotronis) [inline backend/testing/tester.py]: double-check if priority is correct

File: backend/testing/tester.py [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+            if message.delivery_info['routing_key'] == 'update-insert':
+                expected = messages[self.curr_idx]['detection_update_response']
+            elif message.delivery_info['routing_key'] == 'update':
+                expected = messages[self.curr_idx]['detection_hijack_response']
+            elif message.delivery_info['routing_key'] == 'hijack-update':
+                expected = messages[self.curr_idx]['database_hijack_response']
+
+            # compare expected message with received one. exit on mismatch.
+            for key in set(event.keys()).intersection(expected.keys()):
+                if event[key] != expected[key]:
```
Comment (vkotronis) [inline backend/testing/tester.py]: what happens with the ordering of e.g., lists like asns_inf? Shouldn't these be compared in terms of content (e.g., if set(asns_inf_A) == set(asns_inf_B))? Just wondering if we get sth wrong in the expected order, the test will fail. If we require this by design, you can resolve this comment.

Comment (slowr) [reply backend/testing/tester.py]: we make it set comparison as order does not matter
